src/renderers/shared/fiber/ReactFiberCollection.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * Copyright 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactFiberCollection
+ * @flow
+ */
+
+'use strict';
+
+import type {Fiber} from 'ReactFiber';
+
+// A special Map type where the keys are fibers and fiber alternates are treated
+// as equal. Example:
+//
+//   fiberMapSet(fiberMap, fiber, value);
+//   assert(value === fiberMapGet(fiberMap, fiber.alternate));
+//
+// We use arrays instead of native maps because most Map polyfills work by
+// adding a private field to the key type, which causes a de-opt.
+//
+// Right now, we're only using this to keep track of failed error boundaries,
+// where it's expected that the size of collection is small.
+export type FiberMap<T> = {|
+  keys: Array<Fiber>,
+  values: Array<T>,
+|};
+
+// Set version of FiberMap
+export type FiberSet = Array<Fiber>;
+
+// Returns true a and b are equal OR if they are alternates
+function fiberEquals(a: Fiber, b: Fiber) {
+  return a === b || (a.alternate !== null && a.alternate === b);
+}
+
+function createFiberMap() {
+  return {
+    keys: [],
+    values: [],
+  };
+}
+exports.createFiberMap = createFiberMap;
+
+function fiberMapHas<T>(map: FiberMap<T>, fiber: Fiber): boolean {
+  for (let i = 0; i < map.keys.length; i++) {
+    const key = map.keys[i];
+    if (fiberEquals(key, fiber)) {
+      return true;
+    }
+  }
+  return false;
+}
+exports.fiberMapHas = fiberMapHas;
+
+function fiberMapSet<T>(map: FiberMap<T>, fiber: Fiber, value: T) {
+  if (fiberMapHas(map, fiber)) {
+    return;
```

Comment:
If we're not gonna support this properly (by overwriting the value), can we throw?